cli/src/errors/error.rs

```diff
@@ -1,49 +1,92 @@
-#[cfg(backtrace)]
-use std::backtrace::Backtrace;
 use std::convert::From;
+use std::error::Error;
+use std::fmt;
 
 use anyhow::Result;
 use log::error;
-use thiserror::Error;
 
+use crate::errors::ErrorDetails;
 use crate::errors::ExitCode;
 
-#[derive(Error, Debug)]
-#[error("{msg}")]
+#[derive(Debug)]
 pub struct ApolloError {
-    msg: String,
-    #[cfg(backtrace)]
-    backtrace: Backtrace,
-    /// The result of `error.exit_code()`.
-    exit_code: ExitCode,
+    inner: Box<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    kind: ErrorDetails,
+    source: Option<Box<dyn Error>>,
 }
 
 #[allow(dead_code)]
 impl ApolloError {
     /// Returns the process exit code that should be returned if the process exits with this error.
     pub fn exit_code(&self) -> ExitCode {
-        self.exit_code
+        self.inner.kind.exit_code()
+    }
+
+    pub fn from_source<E>(source: E, kind: ErrorDetails) -> Self
+    where
+        E: Into<Box<dyn Error>>,
+    {
+        ApolloError {
+            inner: Box::new(Inner {
+                kind,
+                source: Some(source.into()),
+            }),
+        }
+    }
+
+    pub fn kind(&self) -> &ErrorDetails {
+        &self.inner.kind
     }
 }
 
-/// The failure trait for all Apollo errors.
-pub trait ApolloFail: std::error::Error {
-    /// Returns the process exit code that should be returned if the process exits with this error.
-    fn exit_code(&self) -> ExitCode;
+impl fmt::Display for ApolloError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.inner.kind.fmt(f)
```

Comment:
I [opened a bug](https://github.com/intellij-rust/intellij-rust/issues/5457) for this marking issue with the IJ Rust plugin, but if you wanna make my IntelliJ happy, this is not shown as red:
```rust
(&self.inner.kind).fmt(f)
```

![image](https://user-images.githubusercontent.com/161314/82950449-2bdd0700-9f5a-11ea-8e4f-4bcb619ca47e.png)

vs.

![image](https://user-images.githubusercontent.com/161314/82950462-339cab80-9f5a-11ea-824a-08ceb8d4032d.png)

Reply:
Sure happy to make that change!